vsphere/resource_vsphere_virtual_machine_migrate.go

```diff
@@ -3,35 +3,151 @@ package vsphere
+	vm, err := virtualmachine.FromUUID(client, id)
+	if err != nil {
+		return fmt.Errorf("error fetching virtual machine: %s", err)
+	}
+	props, err := virtualmachine.Properties(vm)
+	if err != nil {
+		return fmt.Errorf("error fetching virtual machine properties: %s", err)
+	}
+
+	// Validate the disks in the VM to make sure that they will work with the new
+	// version of the resource. This is mainly ensuring that all disks are SCSI
+	// disks, but a Read operation is attempted as well to make sure it will
+	// survive that.
+	//
+	// NOTE: This uses the current version of the resource to make this check,
+	// which at some point in time may end up being a higher schema version than
+	// version 2. At this point in time, there is nothing here that would cause
+	// issues (nothing in the sub-resource read logic is reliant on schema
+	// versions), and an empty ResourceData is sent anyway.
+	diskCnt, _ := strconv.Atoi(is.Attributes["disk.#"])
+	maxBus := diskCnt / 15
+	l := object.VirtualDeviceList(props.Config.Hardware.Device)
+	for k, v := range is.Attributes {
+		if !regexp.MustCompile("disk\\.[0-9]+\\.key").MatchString(k) {
+			continue
 		}
-		return is, nil
+		key, _ := strconv.Atoi(v)
+		if key < 1 {
+			continue
+		}
+		device := l.FindByKey(int32(key))
+		if device == nil {
+			continue
+		}
+		ctlr := l.FindByKey(device.GetVirtualDevice().ControllerKey)
+		if ctlr == nil {
+			continue
+		}
+		if sc, ok := ctlr.(types.BaseVirtualSCSIController); ok && sc.GetVirtualSCSIController().BusNumber > int32(maxBus) {
+			maxBus = int(sc.GetVirtualSCSIController().BusNumber)
+		}
+	}
+
+	d := resourceVSphereVirtualMachine().Data(&terraform.InstanceState{})
+	d.Set("scsi_controller_count", maxBus+1)
+	if err := virtualdevice.DiskImportOperation(d, client, object.VirtualDeviceList(props.Config.Hardware.Device)); err != nil {
+		return err
+	}
+
+	rs := resourceVSphereVirtualMachine().Schema
+	var guestNetTimeout string
+	switch is.Attributes["wait_for_guest_net"] {
+	case "false":
+		guestNetTimeout = "-1"
 	default:
-		return is, fmt.Errorf("Unexpected schema version: %d", v)
+		guestNetTimeout = fmt.Sprintf("%v", rs["wait_for_guest_net_timeout"].Default)
 	}
+
+	// The VM should be ready for reading now
+	is.Attributes = make(map[string]string)
+	is.ID = id
+	is.Attributes["imported"] = "true"
+
+	// Set some defaults. This helps possibly prevent diffs where these values
+	// have not been changed.
+	is.Attributes["scsi_controller_count"] = fmt.Sprintf("%v", rs["scsi_controller_count"].Default)
+	is.Attributes["force_power_off"] = fmt.Sprintf("%v", rs["force_power_off"].Default)
+	is.Attributes["migrate_wait_timeout"] = fmt.Sprintf("%v", rs["migrate_wait_timeout"].Default)
+	is.Attributes["shutdown_wait_timeout"] = fmt.Sprintf("%v", rs["shutdown_wait_timeout"].Default)
+	is.Attributes["wait_for_guest_net_timeout"] = guestNetTimeout
+	is.Attributes["scsi_controller_count"] = fmt.Sprintf("%v", maxBus+1)
```

Comment:
In state they won't be. Setting them to their defaults helps clean up the initial plan post-migration (otherwise you get something like `"" => <default value>`)